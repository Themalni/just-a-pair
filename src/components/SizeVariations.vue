<template>
  <div>
    <div class="size-variations pt-2">
      <p class="size-name">Size <span class="accent font-weight-bold text-uppercase">{{ selectedSize }}</span></p>
      <div class="row">
        <div class="col-md-7">
          <div v-for="(size) in sizes" :key="size.index" class="sizes mb-1 waves-effect waves-light" :class="{ 'size-selected': selectedSize === size.number, 'size-disabled': size.available === false }" :disabled="size.available === false" @click="selectShoesSize(size, size.index)">
            <span class="single-size text-uppercase">{{ size.number }}</span>
          </div>
        </div>
        <div class="col-md-5">
          <a href="#sizeGuideModal" class="size-guide-link mb-3" data-toggle="modal" data-target="#sizeGuideModal">
            <md-icon>straighten</md-icon><span class="link-text">Size guide</span>
          </a>
        </div>
      </div>
    </div>
    <size-guide/>
  </div>
</template>

<script>
import SizeGuide from '../components/SizeGuide.vue'

export default {
  name: 'size-variations',
  data() {
    return {
      selectedSize: '36',
      firstAvailableSize: ''
    }
  },
  props: {
    sizes: {
      type: Array
    }
  },
  components: {
    SizeGuide
  },
  methods: {
    selectShoesSize (size) {
      if (size.available === true) {
        return this.selectedSize = size.number
      }
      return size.number
    },
    defaultSize () {
      this.firstAvailableSize = size
    }
  }
}
</script>

<style lang="scss" scoped>
.size-name {
  font-size: 1.2em;

  .accent {
    color: #333;
  }
}
.sizes {
  display: inline-block;
  width: 40px;
  height: 40px;
  background-color: $light-blue;
  margin-right: 0.5em;
  text-align: center;
  font-size: 1.3em;
  color: #fff;
  line-height: 2.2;
  cursor: pointer;
}
.size-selected {
  background-color: $main-blue;
}
.size-disabled {
  background-color: $grey;
  pointer-events: none;
}
.size-guide-link {
  margin-top: 0.5em;
  color: $dark-grey;
  display: block;
  font-size: 1.2em;

  .link-text {
    margin-left: 0.3em;
  }

  &:hover {
    color: $grey;
    text-decoration: none;
  }
}
</style>