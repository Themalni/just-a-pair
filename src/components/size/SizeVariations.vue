<template>
  <div>
    <div class="size-variations pt-2">
      <p class="size-name font-weight-bold">Sizes</p>
      <div class="row">
        <div class="col-md-7">
          <div
            v-for="(size) in sizes"
            :key="size.index"
            class="sizes mb-1 waves-effect waves-light"
            :class="{ 'size-selected': selectedSize === size.number, 'size-disabled': size.available === false }"
            :disabled="size.available === false"
            @click="selectShoesSize(size, size.index)"
          >
            <span class="single-size text-uppercase">{{ size.number }}</span>
          </div>
        </div>
        <div class="col-md-5">
          <a
            href="#sizeGuideModal"
            class="size-guide-link mb-3"
            data-toggle="modal"
            data-target="#sizeGuideModal"
          >
            <md-icon>straighten</md-icon>
            <span class="link-text">Size guide</span>
          </a>
        </div>
      </div>
    </div>
    <size-guide/>
    <transition name="fade">
      <p v-if="selectedSize.length > 0">
        Selected size:
        <span class="accent font-weight-bold">{{ selectedSize }}</span>
      </p>
    </transition>
  </div>
</template>

<script>
import SizeGuide from "./SizeGuide";

export default {
  name: "SizeVariations",
  data() {
    return {
      selectedSize: "",
      firstAvailableSize: ""
    };
  },
  props: {
    sizes: {
      type: Array
    }
  },
  components: {
    SizeGuide
  },
  methods: {
    selectShoesSize(size) {
      this.$emit("selectedSize", size.number);
      if (size.available === true) {
        return (this.selectedSize = size.number);
      }
      return size.number;
    }
  }
};
</script>

<style lang="scss" scoped>
.size-name {
  font-size: 1.4em;

  .accent {
    color: #333;
  }
}
.sizes {
  display: inline-block;
  width: 40px;
  height: 40px;
  background-color: $light-blue;
  margin-right: 0.5em;
  text-align: center;
  font-size: 1.3em;
  color: #fff;
  line-height: 2.2;
  cursor: pointer;
}
.size-selected {
  background-color: $main-blue;
}
.size-disabled {
  background-color: $grey;
  pointer-events: none;
}
.size-guide-link {
  margin-top: 0.5em;
  color: $dark-grey;
  display: block;
  font-size: 1.2em;

  .link-text {
    margin-left: 0.3em;
  }

  &:hover {
    color: $grey;
    text-decoration: none;
  }
}
</style>
