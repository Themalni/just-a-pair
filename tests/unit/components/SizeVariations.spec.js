import { mount, createLocalVue } from '@vue/test-utils';
import SizeVariations from '@/components/Size/SizeVariations';
import mockState from '../utils/mockState';
import Vuex from 'vuex';


const localVue = createLocalVue();
localVue.use(Vuex);

describe('SizeVariations', () => {
    let wrapper
    let store
    let state

    beforeEach(() => {
        state = mockState.state
        store = new Vuex.Store({
            state: state
        })

        wrapper = mount(SizeVariations, {
            store,
            localVue,
            stubs: ['md-icon'],
            propsData: {
                sizes: ['36']
            }
        });
    });

    afterEach(() => {
        wrapper.destroy();
    });

    it('renders correctrly', () => {
        expect(wrapper.html()).toMatchSnapshot();
    })

    it('resets selected size to default value', async () => {
        await wrapper.setData({ selectedSize: '40' })
        expect(wrapper.vm.selectedSize).toBe('40')

        wrapper.vm.resetSize();
        expect(wrapper.vm.selectedSize).toBe(null)
    })

    it('emits selected size and sets `selectedSize` in data', () => {
        wrapper.vm.selectShoesSize('38');
        wrapper.vm.$emit('selectedSize', '38');

        expect(wrapper.emitted().selectedSize).toBeTruthy();
        expect(wrapper.emitted().selectedSize[0]).toEqual(['38']);
        expect(wrapper.vm.selectedSize).toBe('38');
    })

    it('resets selected size when product id has been changed', async () => {
        const resetSizeMock = jest.spyOn(wrapper.vm, 'resetSize')
        store.state.productPreview.id = 999;

        await wrapper.vm.$nextTick();
        expect(resetSizeMock).toHaveBeenCalled();
    })
})