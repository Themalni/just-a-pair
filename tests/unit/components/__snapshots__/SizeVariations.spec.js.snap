// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SizeVariations renders correctrly 1`] = `
<div>
  <div class="size-variations pt-2"><span class="d-block font-weight-bold mt-2 mb-2">Select size</span>
    <div class="row no-gutters">
      <div class="col-md-5">
        <div disabled="disabled" class="sizes mb-1 waves-effect waves-light size-disabled"><span class="single-size text-uppercase"></span></div>
      </div>
      <div class="col"><a href="#sizeGuideModal" data-toggle="modal" data-target="#sizeGuideModal" class="size-guide-link mb-3 ml-3">
          <md-icon-stub>straighten</md-icon-stub> <span class="link-text">Size guide</span>
        </a></div>
    </div>
  </div>
  <div id="sizeGuideModal" tabindex="-1" role="dialog" aria-hidden="true" class="modal fade">
    <div role="document" class="modal-dialog modal-lg">
      <div class="modal-content">
        <div class="modal-header">
          <h5 class="modal-title font-weight-bold">Shoe Size Guide</h5> <button type="button" data-dismiss="modal" aria-label="Close" class="close"><span aria-hidden="true">Ã—</span></button>
        </div>
        <div class="modal-body">
          <h4 class="table-title">Mens footwear</h4>
          <table class="table table-responsive-lg text-center">
            <tbody>
              <tr>
                <th colspan="4" class="main-blue white-text table-category">USA</th>
                <td>7</td>
                <td>7.5</td>
                <td>8</td>
                <td>8.5</td>
                <td>9</td>
                <td>9.5</td>
                <td>10</td>
                <td>10.5</td>
                <td>11</td>
                <td>11.5</td>
                <td>12</td>
                <td>13</td>
              </tr>
              <tr>
                <th colspan="4" class="main-blue white-text table-category">UK</th>
                <td>6</td>
                <td>6.5</td>
                <td>7</td>
                <td>7.5</td>
                <td>8</td>
                <td>8.5</td>
                <td>9</td>
                <td>9.5</td>
                <td>10</td>
                <td>10.5</td>
                <td>11</td>
                <td>12</td>
              </tr>
              <tr>
                <th colspan="4" class="main-blue white-text table-category">EUR</th>
                <td>39.5</td>
                <td>40</td>
                <td>41</td>
                <td>41.5</td>
                <td>42</td>
                <td>42.5</td>
                <td>43</td>
                <td>44</td>
                <td>44.5</td>
                <td>45</td>
                <td>46</td>
                <td>47</td>
              </tr>
              <tr>
                <th colspan="4" class="main-blue white-text">cm</th>
                <td>26</td>
                <td>26.5</td>
                <td>27</td>
                <td>27.5</td>
                <td>28</td>
                <td>28.5</td>
                <td>29</td>
                <td>29.5</td>
                <td>30</td>
                <td>30.5</td>
                <td>31</td>
                <td>31.5</td>
              </tr>
            </tbody>
          </table>
          <h4 class="table-title">Womens footwear</h4>
          <table class="table table-responsive-lg text-center">
            <tbody>
              <tr>
                <th colspan="4" class="main-blue white-text table-category">USA</th>
                <td>6</td>
                <td>6.5</td>
                <td>7</td>
                <td>7.5</td>
                <td>8</td>
                <td>8.5</td>
                <td>9</td>
                <td>9.5</td>
                <td>10</td>
                <td>10.5</td>
                <td>11</td>
              </tr>
              <tr>
                <th colspan="4" class="main-blue white-text table-category">UK</th>
                <td>3.5</td>
                <td>4</td>
                <td>4.5</td>
                <td>5</td>
                <td>5.5</td>
                <td>6</td>
                <td>6.5</td>
                <td>7</td>
                <td>7.5</td>
                <td>8</td>
                <td>8.5</td>
              </tr>
              <tr>
                <th colspan="4" class="main-blue white-text table-category">EUR</th>
                <td>36.5</td>
                <td>37.5</td>
                <td>38</td>
                <td>38.5</td>
                <td>39</td>
                <td>39.5</td>
                <td>40</td>
                <td>41</td>
                <td>41.5</td>
                <td>42</td>
                <td>42.5</td>
              </tr>
              <tr>
                <th colspan="4" class="main-blue white-text">cm</th>
                <td>24</td>
                <td>24.5</td>
                <td>25</td>
                <td>25.5</td>
                <td>26</td>
                <td>26.5</td>
                <td>27</td>
                <td>27.5</td>
                <td>28</td>
                <td>28.5</td>
                <td>29</td>
              </tr>
            </tbody>
          </table>
          <h4 class="table-title">Childrens footwear</h4>
          <table class="table table-responsive-lg text-center">
            <tbody>
              <tr>
                <th colspan="4" class="main-blue white-text table-category">USA</th>
                <td>11</td>
                <td>12</td>
                <td>13</td>
                <td>1</td>
                <td>2</td>
                <td>3</td>
                <td>4</td>
                <td>5</td>
                <td>6</td>
              </tr>
              <tr>
                <th colspan="4" class="main-blue white-text table-category">UK</th>
                <td>10</td>
                <td>11</td>
                <td>12</td>
                <td>13</td>
                <td>1</td>
                <td>2</td>
                <td>3</td>
                <td>4</td>
                <td>5</td>
              </tr>
              <tr>
                <th colspan="4" class="main-blue white-text table-category">EUR</th>
                <td>28</td>
                <td>29.5</td>
                <td>31</td>
                <td>32</td>
                <td>33.5</td>
                <td>35</td>
                <td>36</td>
                <td>37.5</td>
                <td>38.5</td>
              </tr>
              <tr>
                <th colspan="4" class="main-blue white-text">cm</th>
                <td>18.5</td>
                <td>19.5</td>
                <td>20</td>
                <td>21</td>
                <td>22</td>
                <td>23</td>
                <td>23.5</td>
                <td>24.5</td>
                <td>25.5</td>
              </tr>
            </tbody>
          </table>
        </div>
        <div class="modal-footer"><button type="button" data-dismiss="modal" class="btn btn-secondary btn-sm">Close</button></div>
      </div>
    </div>
  </div>
  <transition-stub name="fade">
    <!---->
  </transition-stub>
</div>
`;
